core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractNodeBuildHelper.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    public void addNode(AbstractNode node, Stream_ creator, Stream_ leftParent, Stream_ rightParent) {
+        reversedNodeList.add(node);
+        nodeCreatorMap.put(node, creator);
+        putInsertUpdateRetract(leftParent, TupleLifecycle.ofLeft((LeftTupleLifecycle<? extends AbstractTuple>) node));
+        putInsertUpdateRetract(rightParent, TupleLifecycle.ofRight((RightTupleLifecycle<? extends AbstractTuple>) node));
+    }
+
+    public <Tuple_ extends AbstractTuple> void putInsertUpdateRetract(Stream_ stream,
+            TupleLifecycle<Tuple_> tupleLifecycle) {
+        tupleLifecycleMap.put(stream, tupleLifecycle);
+    }
+
+    public <Tuple_ extends AbstractTuple> void putInsertUpdateRetract(Stream_ stream, List<? extends Stream_> childStreamList,
+            Function<TupleLifecycle<Tuple_>, TupleLifecycle<Tuple_>> tupleLifecycleFunction) {
+        TupleLifecycle<Tuple_> tupleLifecycle = getAggregatedTupleLifecycle(childStreamList);
+        putInsertUpdateRetract(stream, tupleLifecycleFunction.apply(tupleLifecycle));
+    }
+
+    @SuppressWarnings("unchecked")
+    public <Tuple_ extends AbstractTuple> TupleLifecycle<Tuple_>
+            getAggregatedTupleLifecycle(List<? extends Stream_> streamList) {
+        var tupleLifecycles = streamList.stream()
+                .filter(this::isStreamActive)
+                .map(s -> getTupleLifecycle(s, tupleLifecycleMap))
+                .toArray(TupleLifecycle[]::new);
+        return switch (tupleLifecycles.length) {
+            case 0 ->
+                throw new IllegalStateException("Impossible state: None of the streamList (" + streamList + ") are active.");
+            case 1 -> tupleLifecycles[0];
+            default -> TupleLifecycle.aggregate(tupleLifecycles);
+        };
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <Stream_, Tuple_ extends AbstractTuple> TupleLifecycle<Tuple_> getTupleLifecycle(Stream_ stream,
+            Map<Stream_, TupleLifecycle<? extends AbstractTuple>> tupleLifecycleMap) {
+        var tupleLifecycle = (TupleLifecycle<Tuple_>) tupleLifecycleMap.get(stream);
+        if (tupleLifecycle == null) {
+            throw new IllegalStateException("Impossible state: the stream (" + stream + ") hasn't built a node yet.");
+        }
+        return tupleLifecycle;
+    }
+
+    public int reserveTupleStoreIndex(Stream_ tupleSourceStream) {
+        return storeIndexMap.compute(tupleSourceStream, (k, index) -> {
+            if (index == null) {
+                return 0;
+            } else if (index < 0) {
+                throw new IllegalStateException("Impossible state: the tupleSourceStream (" + k
+                        + ") is reserving a store after it has been extracted.");
+            } else {
+                return index + 1;
+            }
+        });
+    }
+
+    public int extractTupleStoreSize(Stream_ tupleSourceStream) {
+        Integer lastIndex = storeIndexMap.put(tupleSourceStream, Integer.MIN_VALUE);
```

Comment:
Should it be zero instead of Integer.MIN_VALUE?

Reply:
Actually, no. This just indicates "no value".
Also, this code did not change from what was there before.